examples/solvers/chambolle_pock_denoising.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Denoising using Chambolle-Pock solver."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+# External
+import numpy as np
+import scipy
+import scipy.ndimage
+import matplotlib.pyplot as plt
+
+# Internal
+import odl
+from odl.solvers import chambolle_pock_solver, f_cc_prox_l2_tv, g_prox_none
+
+
+# Read the image
+im = np.rot90(scipy.misc.lena()[::2, ::2], 3)
+shape = im.shape
+# Rescale max to 1
+im = im / np.max(im[:])
+
+# Discretized spaces
+discr_space = odl.uniform_discr([0, 0], shape, shape)
+
+# Original image
+orig = discr_space.element(im)
+
+# Add noise
+im = im + np.random.normal(0, 0.05, shape)
+# Rescale back to [0, 1]
+im = (im - np.min(im)) / (np.max(im) - np.min(im))
+
+# Data: noisy image
+noisy = discr_space.element(im)
+
+# Gradient operator
+grad = odl.DiscreteGradient(discr_space, method='forward')
+
+# Matrix of operators
+K = odl.ProductSpaceOperator([[odl.IdentityOperator(discr_space)], [grad]])
```

Comment:
> That's not what I meant. It's rather that we show the user in an example how it's supposed to work, which involves the creation of a `ProductSpaceOperator`. That raises the learning curve quite a bit.

Agree.

> My suggestion was rather that the `chambolle_pock_solver` could take the `forward_op` (here the identity) and the `sparsifying_op` separately and, in the first step, generate the `K` internally and go on just as before. That way, users would, in this example, just see the creation of the identity operator and the gradient, without having to think what a product space operator means. That would still encompass all cases we're thinking of, right?

So I read the article now (again). To keep `chambolle_pock_solver` fully general we need to keep this layout it seems, and I think spoiling that would be a misstake. As @aringh  indicated, this would also force us to diverge from the initial paper, which indeed does use a "ProductSpaceOperator". Some examples of a use case would be to solve `Ax = b`, in which case we have no sparsifying_op. Another use case would be to have several sparsifying_op's.

Perhaps we could have a more simple solver that wraps most of these things. Think `chambolle_pock_tv_solver` or similar?

> Regarding the operator norm estimation, I guess I can simply modify the `power_method_opnorm` function such that it accepts sequence-type input and generates the product space operator internally, such that you could call it like this:
    ```python
    K_norm = 1.1 * odl.operator.oputils.power_method_opnorm([ident, grad], 200)
    ```
    I'll look into it.

In this case we can use that `||K|| = sup ||Kx|| / ||x|| = sup ||(Ax, By)|| / ||(x, y)|| = sup (||Ax|| + ||By||) / (||x|| + ||y||) = sup (||A|| ||x|| + ||B|| ||y||) / (||x|| + ||y||) = max(||A||, ||B||)`.

> A simpler example would definitely be fine. But I think we should have an example with an product space operator as well since non-smooth convex problems such as TV are a typical use case for CP.

Correct, perhaps we can have two examples?